### PR TITLE
[ContactStructuralMechanicsApplication] General clean up of the application

### PR DESCRIPTION
**Pull request by loumalouomega**
Additionally fixing some loops that for large simulations give errors in OMP

### CODE REVIEW

**Comment by loumalouomega** [comment]
@philbucher additionally can you take a look if this solves the errors you had when running the tests?

File: applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp [hunk 1]
```diff
@@ -53,22 +53,22 @@ void MasterSlaveProcess::Execute()
...
-            for (IndexType i_node = 0; i_node < this_geometry.size(); ++i_node) {
-                auto& this_node = this_geometry[i_node];
-                if (this_node.IsDefined(INTERFACE)) {
-                    if (!this_node.Is(INTERFACE)) is_interface = false;
-                    else if (!this_node.Is(SLAVE)) is_slave = false;
+            for (IndexType i_node = 0; i_node < r_geometry.size(); ++i_node) {
+                const auto& r_node = r_geometry[i_node];
+                if (r_node.IsDefined(INTERFACE)) {
+                    if (r_node.IsNot(INTERFACE)) is_interface = false;
+                    else if (r_node.IsNot(SLAVE)) is_slave = false;
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: is this threadsafe? I think not

File: applications/ContactStructuralMechanicsApplication/custom_utilities/base_contact_search.cpp [hunk 2]
```diff
@@ -733,18 +708,18 @@ inline void BaseContactSearch<TDim, TNumNodes, TNumNodesMaster>::NotPredefinedMa
...
         if (it_cond->Is(MASTER)) {
             GeometryType& r_geometry = it_cond->GetGeometry();
-            for (auto& node : r_geometry) {
-                node.SetLock();
-                node.Set(MASTER, true);
-                node.UnSetLock();
+            for (NodeType& r_node : r_geometry) {
+                r_node.SetLock();
+                r_node.Set(MASTER, true);
+                r_node.UnSetLock();
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/base_contact_search.cpp:722]: why dont you change this also to atomic?

**Review by philbucher — APPROVED** [review]
Looks ok to me, but I still get the errors (in release):
This one is with multiple threads
~~~
...............ReadMatrixMarketMatrix(): unable to open /home/vicente/src/Kratos/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/A_testing_condensation.mm.
ReadMatrixMarketVector(): unable to open /home/vicente/src/Kratos/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/b_testing_condensation.rhs.
....F................F.........
Ran 46 of 608 test cases in 0.0487895s. 2 failed:
    TestMixedULMLinearSolverThreeDoFSystem Failed with message: 
        Error: Check failed because std::abs(ref_Dx[i] - Dx[i])/std::abs(ref_Dx[i]) = 0.0653955 is not near to 0.0 = 0 within the tolerance 0.05
in /home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp:675:virtual void Testing::TestMixedULMLinearSolverThreeDoFSystem::TestFunction()

    TestShapeFunctionDerivativesQuadrilateral2 Failed with message: 
        Error: YOUR INITIAL SPLITTING DOES NOT COINCIDE WITH THE CURRENT ONE

in /home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_derivatives.cpp:415:void Testing::TestDerivatives(ModelPart&, Condition::Pointer, Condition::Pointer, Condition::Pointer, Condition::Pointer, vector<long unsigned int>, Testing::IndexType, vector<double>, Testing::IndexType, Testing::DerivateToCheck, Testing::CheckLevel) [with long unsigned int TDim = 3ul; long unsigned int TNumNodes = 4ul; Condition::Pointer = shared_ptr<Condition>; Testing::IndexType = long unsigned int]

Unittests: Finished running cpp unit tests! 
Unittests: 
Running python tests ... 
..Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
....F..Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
F...............................................................
======================================================================
FAIL: test_dynamic_search_quad (test_dynamic_search.TestDynamicSearch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_dynamic_search.py", line 161, in test_dynamic_search_quad
    self._dynamic_search_tests(input_filename, 4)
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_dynamic_search.py", line 131, in _dynamic_search_tests
    check.ExecuteFinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/from_json_check_result_process.py", line 157, in ExecuteFinalizeSolutionStep
    self.assertTrue(isclosethis, msg=(str(value) + " != " + str(value_json) + ", rel_tol = " + str(reltol) + ", abs_tol = " + str(tol) + " : Error checking node " + str(node.Id) + " " + variable_name + " results."))
AssertionError: False is not true : 0.27648446766608664 != 0.28343741620962215, rel_tol = 1e-06, abs_tol = 0.001 : Error checking node 1017 NORMAL_GAP results.

======================================================================
FAIL: test_mortar_mapping_quad (test_mortar_mapper.TestMortarMapping)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_mortar_mapper.py", line 210, in test_mortar_mapping_quad
    self._mapper_tests(input_filename, 4, False)
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_mortar_mapper.py", line 176, in _mapper_tests
    check.ExecuteFinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/from_json_check_result_process.py", line 157, in ExecuteFinalizeSolutionStep
    self.assertTrue(isclosethis, msg=(str(value) + " != " + str(value_json) + ", rel_tol = " + str(reltol) + ", abs_tol = " + str(tol) + " : Error checking node " + str(node.Id) + " " + variable_name + " results."))
AssertionError: False is not true : 10.04886216959597 != 10.043189538154905, rel_tol = 1e-06, abs_tol = 0.001 : Error checking node 242 TEMPERATURE results.

----------------------------------------------------------------------
Ran 73 tests in 99.128s
~~~

this is with one thread:
~~~
...............ReadMatrixMarketMatrix(): unable to open /home/vicente/src/Kratos/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/A_testing_condensation.mm.
ReadMatrixMarketVector(): unable to open /home/vicente/src/Kratos/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/b_testing_condensation.rhs.
....F................F.........
Ran 46 of 608 test cases in 0.0399815s. 2 failed:
    TestMixedULMLinearSolverThreeDoFSystem Failed with message: 
        Error: Check failed because std::abs(ref_Dx[i] - Dx[i])/std::abs(ref_Dx[i]) = 0.0515649 is not near to 0.0 = 0 within the tolerance 0.05
in /home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp:675:virtual void Testing::TestMixedULMLinearSolverThreeDoFSystem::TestFunction()

    TestShapeFunctionDerivativesQuadrilateral2 Failed with message: 
        Error: YOUR INITIAL SPLITTING DOES NOT COINCIDE WITH THE CURRENT ONE

in /home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_derivatives.cpp:415:void Testing::TestDerivatives(ModelPart&, Condition::Pointer, Condition::Pointer, Condition::Pointer, Condition::Pointer, vector<long unsigned int>, Testing::IndexType, vector<double>, Testing::IndexType, Testing::DerivateToCheck, Testing::CheckLevel) [with long unsigned int TDim = 3ul; long unsigned int TNumNodes = 4ul; Condition::Pointer = shared_ptr<Condition>; Testing::IndexType = long unsigned int]

Unittests: Finished running cpp unit tests! 
Unittests: 
Running python tests ... 
..Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
....F..Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
Quadrilateral3D4: detJ =	2.53952e-07 DeltaX =	[2](401.4,-512.91) stopping calculation. Iteration:	383
F.......ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
F.......................................................
======================================================================
FAIL: test_dynamic_search_quad (test_dynamic_search.TestDynamicSearch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_dynamic_search.py", line 161, in test_dynamic_search_quad
    self._dynamic_search_tests(input_filename, 4)
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_dynamic_search.py", line 131, in _dynamic_search_tests
    check.ExecuteFinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/from_json_check_result_process.py", line 157, in ExecuteFinalizeSolutionStep
    self.assertTrue(isclosethis, msg=(str(value) + " != " + str(value_json) + ", rel_tol = " + str(reltol) + ", abs_tol = " + str(tol) + " : Error checking node " + str(node.Id) + " " + variable_name + " results."))
AssertionError: False is not true : 0.27648446766608664 != 0.28343741620962215, rel_tol = 1e-06, abs_tol = 0.001 : Error checking node 1017 NORMAL_GAP results.

======================================================================
FAIL: test_mortar_mapping_quad (test_mortar_mapper.TestMortarMapping)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_mortar_mapper.py", line 210, in test_mortar_mapping_quad
    self._mapper_tests(input_filename, 4, False)
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/test_mortar_mapper.py", line 176, in _mapper_tests
    check.ExecuteFinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/from_json_check_result_process.py", line 157, in ExecuteFinalizeSolutionStep
    self.assertTrue(isclosethis, msg=(str(value) + " != " + str(value_json) + ", rel_tol = " + str(reltol) + ", abs_tol = " + str(tol) + " : Error checking node " + str(node.Id) + " " + variable_name + " results."))
AssertionError: False is not true : 10.04886216959597 != 10.043189538154905, rel_tol = 1e-06, abs_tol = 0.001 : Error checking node 242 TEMPERATURE results.

======================================================================
FAIL: test_execution (SmallTests.SimplePatchTestThreeDMeshTying)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/philippb/software/Kratos_Master/applications/ContactStructuralMechanicsApplication/tests/contact_structural_mechanics_test_factory.py", line 78, in test_execution
    self.test.RunSolutionLoop()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/analysis_stage.py", line 60, in RunSolutionLoop
    self.FinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/analysis_stage.py", line 135, in FinalizeSolutionStep
    process.ExecuteFinalizeSolutionStep()
  File "/home/philippb/software/Kratos_Master/kratos/python_scripts/from_json_check_result_process.py", line 157, in ExecuteFinalizeSolutionStep
    self.assertTrue(isclosethis, msg=(str(value) + " != " + str(value_json) + ", rel_tol = " + str(reltol) + ", abs_tol = " + str(tol) + " : Error checking node " + str(node.Id) + " " + variable_name + " results."))
AssertionError: False is not true : -100033387275.22505 != -0.04163056265274044, rel_tol = 1e-06, abs_tol = 0.001 : Error checking node 195 DISPLACEMENT_Y results.

----------------------------------------------------------------------
Ran 73 tests in 94.467s
~~~

to solve the c-issue with the path I think you have to do sth in cmake

File: applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: I will check, in any case is not used always

File: applications/ContactStructuralMechanicsApplication/custom_utilities/base_contact_search.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/base_contact_search.cpp:722]: Doesn't work

File: applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: I checked, it is, the node is const, the parallel region is only for conditions

Comment (philbucher) [reply applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: thats the point
if the conditions share a node, then the node might be accessed twice from two threads

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: Yeah, but accessing is threadsafe, writing is not, right @roigcarlo ?

Comment (philbucher) [reply applications/ContactStructuralMechanicsApplication/custom_processes/master_slave_process.cpp:66]: ah sorry you are only accessing, confused myself.
Nevermind, sorry for the noise

**Pull request merged**